src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
-    }
-    const container = part.startNode.parentNode as HTMLElement | ShadowRoot |
-        DocumentFragment;
-    let index = -1;
-    let currentMarker = part.startNode.nextSibling!;
+  return directive((directivePart: NodePart): void => {
+    // Old part list & map is retrieved from the last render (associated with
+    // the part created for this instance of the directive)
+    let oldParts = partListCache.get(directivePart) || [];
+    let oldKeyToIndexMap = partMapCache.get(directivePart) || new Map();
 
-    for (const item of items) {
-      let result;
-      let key;
-      try {
-        ++index;
-        result = template !(item, index);
-        key = keyFn ? keyFn(item) : index;
-      } catch (e) {
-        console.error(e);
-        continue;
-      }
+    // New part list will be built up as we go (either reused from old parts or
+    // created for new keys in this update). This is saved in partListCache
+    // at the end of the update.
+    const newParts: Array<KeyedNodePart> = [];
+    const newKeyToIndexMap: Map<any, number> = new Map();
 
-      // Try to reuse a part
-      let itemPart = keyMap.get(key);
-      if (itemPart === undefined) {
-        // TODO(justinfagnani): We really want to avoid manual marker creation
-        // here and instead use something like part.insertBeforePart(). This
-        // requires a little refactoring, like iterating through values and
-        // existing parts like NodePart#_setIterable does. We can also remove
-        // keyMapCache and use part._value instead.
-        // But... repeat() is badly in need of rewriting, so we'll do this for
-        // now and revisit soon.
-        const marker = createMarker();
-        const endNode = createMarker();
-        container.insertBefore(marker, currentMarker);
-        container.insertBefore(endNode, currentMarker);
-        itemPart = new NodePart(part.templateFactory);
-        itemPart.insertAfterNode(marker);
-        if (key !== undefined) {
-          keyMap.set(key, itemPart);
-        }
-      } else if (currentMarker !== itemPart.startNode) {
-        // Existing part in the wrong position
-        const end = itemPart.endNode.nextSibling!;
-        if (currentMarker !== end) {
-          reparentNodes(container, itemPart.startNode, end, currentMarker);
-        }
+    // New result list is eagerly generated from items and marked with its key.
+    // This is saved in partMapCache at the end of the update.
+    const newResults: Array<KeyedTemplateResult> = [];
+    let index = 0;
+    for (let item of items) {
+      const key = keyFn ? keyFn(item) : index;
+      if (newKeyToIndexMap.has(key)) {
+        console.warn(`Duplicate key '${
+            key}' detected in repeat; behavior is undefined.`);
       } else {
-        // else part is in the correct position already
-        currentMarker = itemPart.endNode.nextSibling!;
+        const result = newResults[index] = template !(item, index);
```

Comment:
Arr, if duplicate keys are detected the `index` passed here should still reflect the `items` index, not the final result/part index... worth fixing?

Reply:
The actual "undefined" behavior is that only the first item for a key with duplicates will be included in the results; the rest will be skipped. But to provide the correct index to the template function, we need to increment `index` for each item regardless; however since we also store the index in the key map, that one needs to reflect the index taking into account skips -- which is just `newKeyToIndexMap.size`